### PR TITLE
[IRGen][AArch64][RISCV] Generalize bitcast between i1 predicate vector and i8 fixed vector.

### PR DESCRIPTION
**Pull request by topperc**
Instead of only handling vscale x 16 x i1 predicate vectors, handle any scalable i1 vector where the known minimum is divisible by 8.

This will be used on RISC-V where we have multiple sizes of predicate types.

Though I do wonder if we have the bitcast on the wrong side. Maybe we should be casting the fixed type to an i1 fixed vector? That would better handle cases where the mininum elements isn't a multiple of 8, but vscale*min_elements is.

### CODE REVIEW

**Comment by topperc** [comment]
Ping

File: clang/lib/CodeGen/CGExprScalar.cpp [hunk 1]
```diff
@@ -2136,14 +2136,16 @@ Value *ScalarExprEmitter::VisitCastExpr(CastExpr *CE) {
...
-        // If we are casting a fixed i8 vector to a scalable 16 x i1 predicate
+        // If we are casting a fixed i8 vector to a scalable i1 predicate
         // vector, use a vector insert and bitcast the result.
         bool NeedsBitCast = false;
-        auto PredType = llvm::ScalableVectorType::get(Builder.getInt1Ty(), 16);
         llvm::Type *OrigType = DstTy;
-        if (ScalableDst == PredType &&
-            FixedSrc->getElementType() == Builder.getInt8Ty()) {
-          DstTy = llvm::ScalableVectorType::get(Builder.getInt8Ty(), 2);
+        if (ScalableDst->getElementType()->isIntegerTy(1) &&
```
Comment (sdesmalen-arm) [inline clang/lib/CodeGen/CGExprScalar.cpp:2143]: nit: While you're touching this code, maybe refactor this whole block to remove the control flow by `NeedsBitCast`, by something like this:

```
if (const auto *FixedSrc = dyn_cast<llvm::FixedVectorType>(SrcTy)) {
  if (const auto *ScalableDst = dyn_cast<llvm::ScalableVectorType>(DstTy)) {
    llvm::Value *Zero = llvm::Constant::getNullValue(CGF.CGM.Int64Ty);
    if (ScalableDst->getElementType()->isIntegerTy(1) &&
        ScalableDst->getElementCount().isKnownMultipleOf(8) &&
        FixedSrc->getElementType()->isIntegerTy(8)) {
      auto InsertTy = llvm::ScalableVectorType::get(
          FixedSrc->getElementType(), FixedSrc->getNumElements());
      llvm::Value *Insert = Builder.CreateInsertVector(
          InsertTy, llvm::UndefValue::get(InsertTy), Src, Zero,
          "cast.scalable");
      return Builder.CreateBitCast(Insert, DstTy);
    } else if (FixedSrc->getElementType() == ScalableDst->getElementType())
      return Builder.CreateInsertVector(DstTy, llvm::UndefValue::get(DstTy),
                                        Src, Zero, "cast.scalable");
  }
}
```

What do you think?

(similar suggestion for the other case below for scalable -> fixed)

**Review by sdesmalen-arm — COMMENTED** [review]
Generalising this code makes sense, 16 should never have been hardcoded here.

Is it possible to add a test for the case where the predicate type is not `<vscale x 16 x i1>`?

> Though I do wonder if we have the bitcast on the wrong side. Maybe we should be casting the fixed type to an i1 fixed vector? That would better handle cases where the mininum elements isn't a multiple of 8, but vscale*min_elements is.

I'm not sure if there are any fundamental reasons not to do this, but at least for AArch64 `16 x i1` is not a legal type which leads to some inefficient legalisation code. I can see however that the IR is missing some folds, so perhaps with those folds added making this change is fine.

**Comment by topperc** [comment]
> Generalising this code makes sense, 16 should never have been hardcoded here.
> 
> Is it possible to add a test for the case where the predicate type is not `<vscale x 16 x i1>`?

I rebased, which picked up more tests that are affected for RISC-V. This also pointed out that I missed very similar code in CGCall.cpp so that's been updated now.

File: clang/lib/CodeGen/CGExprScalar.cpp (hunk 1 above)
Comment (topperc) [reply clang/lib/CodeGen/CGExprScalar.cpp:2146]: I plan to do another update to address this.

File: clang/test/CodeGen/attr-riscv-rvv-vector-bits-call.c [hunk 2]
```diff
@@ -70,13 +70,17 @@ fixed_float64m1_t call_float64_ff(fixed_float64m1_t op1, fixed_float64m1_t op2)
...
 // CHECK-LABEL: @call_bool1_ff(
 // CHECK-NEXT:  entry:
-// CHECK-NEXT:    [[SAVED_VALUE4:%.*]] = alloca <vscale x 64 x i1>, align 8
-// CHECK-NEXT:    [[RETVAL_COERCE:%.*]] = alloca <vscale x 64 x i1>, align 8
-// CHECK-NEXT:    [[TMP0:%.*]] = tail call <vscale x 64 x i1> @llvm.riscv.vmand.nxv64i1.i64(<vscale x 64 x i1> [[OP1_COERCE:%.*]], <vscale x 64 x i1> [[OP2_COERCE:%.*]], i64 256)
-// CHECK-NEXT:    store <vscale x 64 x i1> [[TMP0]], ptr [[SAVED_VALUE4]], align 8, !tbaa [[TBAA4:![0-9]+]]
-// CHECK-NEXT:    [[TMP1:%.*]] = load <32 x i8>, ptr [[SAVED_VALUE4]], align 8, !tbaa [[TBAA8:![0-9]+]]
-// CHECK-NEXT:    store <32 x i8> [[TMP1]], ptr [[RETVAL_COERCE]], align 8
-// CHECK-NEXT:    [[TMP2:%.*]] = load <vscale x 64 x i1>, ptr [[RETVAL_COERCE]], align 8
+// CHECK-NEXT:    [[OP1:%.*]] = alloca <32 x i8>, align 8
```
Comment (topperc) [inline clang/test/CodeGen/attr-riscv-rvv-vector-bits-call.c:73]: Need to check what happened here.

File: clang/lib/CodeGen/CGExprScalar.cpp (hunk 1 above)
Comment (topperc) [reply clang/lib/CodeGen/CGExprScalar.cpp]: I kept the control flow, but removed the `NeedsBitCast` variable. We can compare types to see if the cast is needed.

**Review by sdesmalen-arm — APPROVED** [review]

**Pull request merged**